### PR TITLE
Framebuffer fixes

### PR DESCRIPTION
**Pull request by ibgreen**
* Prevent `AnimationLoop` from creating unused framebuffer
* Add `defaultFramebuffer` object, allowing the default framebuffer to be manipulated using the `Framebuffer` class methods.
* Add `readDataUrl` and `readImage` methods, read out texture to HTML Image
* Fixes to `blit` and `getAttachmentParameter`

### CODE REVIEW

File: src/webgl/framebuffer.js [hunk 1]
```diff
@@ -241,14 +265,20 @@ export default class Framebuffer extends Resource {
...
-    pixelArray = null
+    type, // Auto deduced from pixelArray or gl.UNSIGNED_BYTE
+    pixelArray = null,
+    attachment = GL_COLOR_ATTACHMENT0 // TODO
   }) {
     const {gl} = this;
 
+    // TODO - Move into readPixels
+    if (attachment === GL.COLOR_ATTACHMENT0 && this.handle === null) {
+      attachment = GL.FRONT;
```
Comment (1chandu) [inline src/webgl/framebuffer.js]: > attachment = GL.FRONT;

`attachment` seem to be not used any where, in default case, following `this.bind()` is going to bund `null` handle any way, why do we need to handle this separately ?

File: src/webgl/framebuffer.js [hunk 2]
```diff
@@ -269,6 +299,40 @@ export default class Framebuffer extends Resource {
...
+    return canvas.toDataURL();
+  }
+
+  // Reads pixels into an HTML Image
+  readImage({
+    attachment = GL.COLOR_ATTACHMENT0,
+    image
+  } = {}) {
+    /* global Image */
+    const dataUrl = this.createDataUrl();
```
Comment (1chandu) [inline src/webgl/framebuffer.js]: > this.createDataUrl()

Where is this defined? not in super class too.

File: src/webgl/framebuffer.js [hunk 3]
```diff
@@ -269,6 +299,40 @@ export default class Framebuffer extends Resource {
     return pixelArray;
   }
 
+  // Reads pixels as a dataUrl
+  readDataUrl({
+    attachment = GL.COLOR_ATTACHMENT0
```
Comment (1chandu) [inline src/webgl/framebuffer.js]: `attachment` not used any where.

File: src/webgl/framebuffer.js [hunk 4]
```diff
@@ -269,6 +299,40 @@ export default class Framebuffer extends Resource {
...
+    const imageData = context.createImageData(width, height);
+    imageData.data.set(data);
+    context.putImageData(imageData, 0, 0);
+
+    return canvas.toDataURL();
+  }
+
+  // Reads pixels into an HTML Image
+  readImage({
+    attachment = GL.COLOR_ATTACHMENT0,
```
Comment (1chandu) [inline src/webgl/framebuffer.js]: same, `attachment` not used any where.

File: src/webgl/framebuffer.js [hunk 5]
```diff
@@ -348,16 +412,17 @@ export default class Framebuffer extends Resource {
...
+
     gl.bindFramebuffer(GL_READ_FRAMEBUFFER, srcFramebuffer.handle);
     gl.bindFramebuffer(GL_DRAW_FRAMEBUFFER, this.handle);
-    gl.blitFramebuffer(
-      srcX0, srcY0, srcX1, srcY1,
-      dstX0, dstY0, dstX1, dstY1,
-      mask,
-      filter
-    );
+    gl.blitFramebuffer(srcX0, srcY0, srcX1, srcY1, dstX0, dstY0, dstX1, dstY1, mask, filter);
```
Comment (1chandu) [inline src/webgl/framebuffer.js]: Nit: assert or warn message when 
1. mask === 0
2. srcFramebuffer.handle === this.handle

File: src/webgl/framebuffer.js [hunk 6]
```diff
@@ -388,14 +453,16 @@ export default class Framebuffer extends Resource {
...
     attachment = GL_COLOR_ATTACHMENT0,
     pname
   } = {}) {
-    const fallback = this._getAttachmentParameterFallback(pname);
-    return fallback !== null ?
-      fallback :
-      this.gl.getFramebufferAttachmentParameter(target, attachment, pname);
+    let value = this._getAttachmentParameterFallback(pname);
+    if (value === null) {
+      this.gl.bindTexture(GL_FRAMEBUFFER, this.handle);
```
Comment (1chandu) [inline src/webgl/framebuffer.js]: Nit: update comments
`passed pname given the target and attachment.` -> `passed pname of given attachment for 'FRAMEBUFFER' target`

Comment (ibgreen) [reply src/webgl/framebuffer.js]: > passed pname given the target and attachment.
Updated

**Pull request merged**